### ISSUE TITLE
Supporting empty vs nil arrays in canonical go structs

### ISSUE THREAD

**Issue by smarterclayton** [gogo/protobuf#181]
I'd like to be able to have Go code

```
type Foo struct {
  A []string
}
```

generated from a proto2 IDL where `A`, if serialized as nil, is still nil when deserialized, and if empty, is still empty when deserialized.  I'd like that to be aligned with proto3 and only_one, which requires an intermediate message type.  I would like to propose (and implement) something along the following:

```
// proto2
message Foo {
  optional StringSlice A = 1
}

message StringSlice {
  option (gogoproto.inline) = true;
  repeated string B = 1
}
```

that results in a generated Go struct and unmarshaller for 

```
type Foo {
  A []string
}
```

where a missing StringSlice results in []string being nil, and a present but empty StringSlice results in an empty []string.

The option is just a suggestion - I wouldn't mind applying nullable=false to the parent but that might not be backwards compatible for existing structs.

In proto3 the same optimization would be applied to a union that has only a string array and the null type.

Thoughts?

**Comment by smarterclayton** [comment]
@wojtek-t re: kube protobuf for detecting empty strings

**Comment by awalterschulze** [comment]
What about embed?

```proto
// proto2
message Foo {
  optional StringSlice A = 1 [(gogoproto.embed)=true];
}

message StringSlice {
  repeated string B = 1
}
```

That should generate

```go
type Foo struct {
  *StringSlice
}

type StringSlice {
  A []string
}
```

What do you think?

**Comment by smarterclayton** [comment]
I like that.

**Comment by awalterschulze** [comment]
Ok well that feature is done you can just use it :)

**Comment by smarterclayton** [comment]
:)

**Comment by awalterschulze** [comment]
Can I close the issue?

**Comment by smarterclayton** [comment]
Hrm, I'm not sure that is actually the use case I need to solve.  Our case is I'm trying to turn the nested struct into a []string so it's equivalent.

**Comment by awalterschulze** [comment]
Yes is not exactly the same.  You could also use this with nullable=false then you can get

```go
type Foo struct {
  StringSlice
}

type StringSlice {
  A []string
}
```

Is this better?

**Comment by smarterclayton** [comment]
Sorry, was not reading clearly - so we're trying to generate IDL from existing structs and be able to get back the existing structs.  I'd *prefer* to have the ability to have a []string represent both empty and nil in the Go struct, but have the protobuf message match what is necessary to carry the info on the wire.  Embed is possible, but it is less nice because it involves us changing our internal structs anyway.  If it's onerous to do the transformation (i.e., doesn't feel like something that should be in gogo proto) that's fine, I just wanted to approach it.

Underlying philosophy is we have idiomatic Go structs, and are generating IDL, and then want to be able to leverage gogo to generate back to the exact same idiomatic go structs.  That's not quite the gogo mission, but it does at least align with the theory of making protobuf generate idiomatic go.

**Comment by smarterclayton** [comment]
Extending this to the future, when we have one_of types, won't we want a way to collapse a message that may be one_of Null into a canonical go struct?

I.e. if I have a bunch of protos that can be `one_of { message, nullable }`, won't I want a convenient way to reduce that to `pointer to struct defined by message`?

**Comment by awalterschulze** [comment]
Ok I think I am starting to understand. You don't simply want to be able to generate an idiomatic go struct, but you also want to change the marshaling and unmarshaling semantics.
So an empty slice for a repeated field would simply not send any data over the wire and result in a resulting nil slice when unmarshaling.
You want to somehow preserve the emptiness of a repeated field over the wire.
Am I understanding correctly?

**Comment by smarterclayton** [comment]
Yes.  In JSON the distinction between null and [] is valid (and relevant for some objects), and as we're trying to keep the logical structure in go the same we end up having to duplicate it.

I "hacked" (it's the only word that describes it) around for now with a custom generator that uses a type alias `type Foo []string` with special semantics during generation to create a message with a single `Items []string` child.  And I'm doing some AST transforms of the marshaller/unmarshaller to preserve the distinction.  

Having just come back to this, it's somewhat equivalent to inlining a single field in a message into a Go struct field (for proto2) and similar to `one_of { message, nullable }`.

**Comment by awalterschulze** [comment]
I am struggling to understand your hack.  Does your Foo type have its own marshal method?

Interesting idea with oneof.
But how exactly will this oneof type look, maybe you could describe it more detail?

**Comment by awalterschulze** [comment]
hello?

**Comment by smarterclayton** [comment]
Example was:

Canonical type: https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/authorization/v1/types.go#L147

generates to proto:
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/authorization/v1/generated.proto#L37

which then is able to distinguish between nil and empty (if the message is not-nil, the slice is considered empty if it has length zero).

**Comment by awalterschulze** [comment]
Ok so this already works ?

**Comment by jianzhangbjz** [comment]
Already works? Here is my encountered problem: [#46024](https://github.com/kubernetes/kubernetes/issues/46024) how to fix it?

**Comment by jianzhangbjz** [comment]
Referenced to [add tokenreviews endpoint to implement webhook ](https://github.com/kubernetes/kubernetes/pull/28788), solved, sorry for bothering.

**Comment by awalterschulze** [comment]
no problem :)

**Comment by gpopovic** [comment]
@smarterclayton did you find a solution for this?
